### PR TITLE
[M1092] Ingest and validation changes

### PR DESCRIPTION
**Pull request by gridcell**
## Release Notes

- **New Features**
  - Added bulk validation handling in the ingest process, improving efficiency and error management.
  - Introduced new validation methods and constraints across various serializers, enhancing data integrity.
  - Added support for deferred validation execution in the validation process.

- **Bug Fixes**
  - Improved error handling for input validation in multiple serializers.

- **Enhancements**
  - Expanded serializer capabilities with new fields related to fish species and management data.
  - Updated existing fields with validation constraints to ensure data accuracy.

### CODE REVIEW

**Comment by gridcell** [comment]
[bulk validation and submission](https://trello.com/c/aMUk1mIM/1092-bulk-validation-and-submission)

File: src/api/resources/sample_units_base.py [hunk 1]
```diff
@@ -21,6 +21,8 @@ class SampleUnitSerializer(BaseAPISerializer):
         max_digits=3,
         decimal_places=1,
         coerce_to_string=False,
+        min_value=0,
```
Comment (saanobhaai) [inline src/api/resources/sample_units_base.py:24]: coderabbit is right -- we should change src/api/submission/protocol_validations.py to be 0,40.
And here as elsewhere, remove each model field validator for which we're adding a serializer field validator.

File: src/api/resources/quadrat_collection.py [hunk 2]
```diff
@@ -1,9 +1,19 @@
...
 from .base import BaseProjectApiViewSet
 from .sample_units_base import SampleUnitFilterSet, SampleUnitSerializer
 
 
 class QuadratCollectionSerializer(SampleUnitSerializer):
+    quadrat_size = serializers.DecimalField(
+        max_digits=6,
+        decimal_places=2,
+        coerce_to_string=False,
+        min_value=0,
```
Comment (saanobhaai) [inline src/api/resources/quadrat_collection.py:13]: same >0 thing

File: src/api/models/__init__.py [hunk 3]
```diff
@@ -8,6 +8,7 @@
 from .base import Country  # noqa: F401
 from .base import JSONMixin  # noqa: F401
 from .base import Profile  # noqa: F401
+from .base import validate_max_year  # noqa: F401
```
Comment (saanobhaai) [inline src/api/models/__init__.py:11]: I think if we're moving field validation out of models, we should make that clear by removing field validators and moving functions like this into utils or resources.

File: src/api/resources/management.py [hunk 4]
```diff
@@ -53,7 +53,12 @@ class ManagementSerializer(BaseAPISerializer):
     rules = serializers.SerializerMethodField(source="get_rules")
     project_name = serializers.SerializerMethodField()
     size = serializers.DecimalField(
-        max_digits=12, decimal_places=3, coerce_to_string=False, allow_null=True, required=False
+        max_digits=12,
+        decimal_places=3,
+        coerce_to_string=False,
+        allow_null=True,
+        required=False,
+        min_value=0,
```
Comment (saanobhaai) [inline src/api/resources/management.py:61]: >0

File: src/api/resources/fish_species.py [hunk 5]
```diff
@@ -39,12 +39,41 @@ class FishSpeciesSerializer(M2MSerializerMixin, CreateOrUpdateSerializerMixin, B
...
+    )
+    max_length = serializers.DecimalField(
+        max_digits=6,
+        decimal_places=2,
+        coerce_to_string=False,
+        required=False,
+        allow_null=True,
+        min_value=1,
+        max_value=2000,
+    )
```
[the remark this answers is not in the document]
Comment (saanobhaai) [reply src/api/resources/fish_species.py:59]: max is set to much higher to accommodate occasional sharks.

File: src/api/resources/collect_record.py [hunk 6]
```diff
@@ -179,6 +180,8 @@ def ingest(self, request, project_pk, *args, **kwargs):
         if "errors" in ingest_output:
             errors = ingest_output["errors"]
             return Response(errors, status=400)
+        elif "validate" in ingest_output:
+            return Response(ingest_output["validate"], status=400)
```
Comment (saanobhaai) [inline src/api/resources/collect_record.py:184]: Should probably make format of this and previous `errors` match

Comment (gridcell) [reply src/api/resources/collect_record.py:184]: @saanobhaai Wondering what that would look like?

File: src/api/resources/sampleunitmethods/beltfishmethod.py [hunk 7]
```diff
@@ -72,7 +72,7 @@ class Meta:
         }
 
     def validate_size(self, value):
-        if value is not None and value < 0:
+        if value is not None and value <= 0:
             raise serializers.ValidationError("Size must be greater than or equal to 0.")
```
Comment (saanobhaai) [inline src/api/resources/sampleunitmethods/beltfishmethod.py:76]: message should not include `or equal to`

File: src/api/resources/collect_record.py (hunk 6 above)
Comment (saanobhaai) [reply src/api/resources/collect_record.py:184]: Sorry I meant the code, not the return vals, i.e. in one case we're setting a var (`errors`) and in the other we're not and using `ingest_output` directly.

**Review by saanobhaai — CHANGES_REQUESTED** [review]
Two tiny items left...

**Review by saanobhaai — APPROVED** [review]

**Pull request merged**